staging/src/k8s.io/api/resource/v1alpha3/types.go

```diff
@@ -532,10 +532,34 @@ type CELDeviceSelector struct {
 	//
 	//     cel.bind(dra, device.attributes["dra.example.com"], dra.someBool && dra.anotherBool)
 	//
+	// The length of the expression must be smaller or equal to 10 Ki. The
+	// cost of evaluating it is also limited based on the estimated number
+	// of logical steps. Validation against those limits happens only when
+	// setting an expression for the first time or when changing
+	// it. Therefore it is possible to change these limits without
+	// affecting stored expressions. Those remain valid.
```

Comment:
```suggestion
	// it.  If the limits are changed in a future Kubernetes release, existing users are guaranteed that existing expressions will continue to be valid.
```

(To make it clear that there is no way for Kubernetes users to change the limits)

Reply:
Applied.

Reply:
I started to wonder about this... Suppose some stored expression passed the cost limit check in 1.32 and in 1.33 we reduce the limit. The expression is still valid (as stated here and implemented in the validation code), but the same reduced limit is also used as runtime limit with `cel.CostLimit(resourceapi.CELSelectorExpressionMaxCost)`.

Doesn't that mean that a valid expression then can fail to get evaluated at runtime?

Is that runtime cost check really providing some additional value? Either the cost estimation is correct, then the runtime check isn't needed, or the estimation is just a guess, then perhaps it shouldn't be used for validation because it gives the wrong impression of "this expression is fine"?

Reply:
There's enough CEL now that a writeup (blog article ?) for https://k8s.dev/ would help. My opinion anyway.

Reply:
> So, in practice we've been very careful not to change cost on users.

So we can increase it (with care, like considering downgrade scenarios), but not reduce it. I thought we could be more flexible here, but not if we then also apply a runtime check.

>  > Is that runtime cost check really providing some additional value?
>
> Having a way to halt execution at runtime seems pragmatic.

Doing that based on context cancellation makes sense. If evaluation times out, it tells admins that they need a faster machine or that they have to tweak those timeouts. This is entirely theoretic for DRA though, because there are no such limits at the moment.

But I am still not sure why DRA CEL should enable the additional runtime cost check. Can I use `cel.CostLimit` to disable the default cost limit checking? Then I would get the behavior described above ("existing users are guaranteed that existing expressions will continue to be valid")  plus "won't fail at runtime", which I find useful.

It gives us a bit more freedom to change this limit if that turns out to be useful.

> We typically don't say anything in the API docs about changing cost, so it might be the best to to the same here 

Agreed, it doesn't need to be in the user-facing OpenAPI spec. But I want it on the `CELSelectorExpressionMaxCost` constant as guidance for what changes are and what aren't allowed.

Reply:
> Agreed, it doesn't need to be in the user-facing OpenAPI spec. But I want it on the CELSelectorExpressionMaxCost constant as guidance for what changes are and what aren't allowed.

Changed. I'll squash that when we are done with the review.

/hold

> But I am still not sure why DRA CEL should enable the additional runtime cost check.

I went ahead and disabled that runtime cost check in an additional commit because that makes more sense to me. Please ~~yell at me~~ politely point out why I am mistaken here :grin: 

I also modified one unit test case such that it triggers the runtime cost check before that commit and no longer triggers with it.

Reply:
I added a benchmark because I was curious whether I would see a difference when I hack out the cost tracking at runtime. I got some (according to benchstat) statistically relevant improvements, but not enough to make further changes.

The important aspect is that clients should trust the cost estimate in the apiserver validation, which is in this PR now.

Reply:
I'd prefer we keep runtime cost in place.

My perspective comes from having worked on the internals of CEL to enable the cost systems, and I trust runtime cost more than I trust estimated cost.  The logic is simpler and more "obviously correct" than estimated cost.  I'm worried something might slip through if we don't have a runtime limit.

cc @cici37, @liggitt

Reply:
So we end up with the "validation accepted a CEL expression, runtime disagrees and aborts" error scenario.

I am less worried about evaluation taking longer than expected for DRA because scheduling times are already pretty fluid, depending on how complex the problem is. But I'm okay with keeping the runtime limit. I can't update the code right away (on a short vacation this week), but will do so soon.

The comment about the cost limit then becomes something along the lines of "the limit can be changed because stored expressions remain valid" but with the caveat that "clients, in particular the scheduler, must be recompiled to use the same modified limit as the apiserver", which affects version skew scenarios.

Reply:
| So we end up with the "validation accepted a CEL expression, runtime disagrees and aborts" error scenario.

True, but only if estimated cost is wrong (that is, it's bugged), since the guarantee is min(estimated cost) <= runtime cost <= max(estimated cost) when everything is working as intended.

| I can't update the code right away (on a short vacation this week), but will do so soon.

Sounds good. 

| The comment about the cost limit then becomes something along the lines of "the limit can be changed because stored expressions remain valid" but with the caveat that "clients, in particular the scheduler, must be recompiled to use the same modified limit as the apiserver", which affects version skew scenarios.

Another option would be to just not say anything changing limits here.  If we do change them, we can document the situation at that time.  I think it's great you've though the situation through though... having the CEL expressions run in the scheduler does complicate this situation.

Reply:
I've restored the runtime limit in my recent push: https://github.com/kubernetes/kubernetes/compare/6ad744da6d14bc149930eb9ad0a715aad0c2dd40..d53cb79cec11094099205a8c114a416b430a6c2c

> Another option would be to just not say anything changing limits here.

These constants don't show up in the Kubernetes API docs for end users. They are there for my fellow developers who may have to continue maintaining DRA when you or I are not available to answer questions. So capturing our discussion now in the code IMHO is useful.

> having the CEL expressions run in the scheduler does complicate this situation

Indeed. I would still prefer to not do that, but let's proceed with it :sweat_smile:

Reply:
> Changed. I'll squash that when we are done with the review.
> 
> /hold

Commit history was already cleaned before the recent LGTMs.

/hold cancel